flight/Modules/System/systemmod.c

```diff
@@ -310,32 +399,92 @@ static void systemPeriodicCb(UAVObjEvent *ev, void *ctx, void *obj_data, int len
 	TaskMonitorUpdateAll();
 #endif
 
+#if defined(PIOS_INCLUDE_ANNUNC)
+	// Figure out what we should be doing.
+
+	static const char *blink_string = NULL;
+	static uint32_t blink_state = 0;
+	static uint8_t blink_prio = 0;
+	static bool ever_armed = false;
+	static bool is_manual_control = false;
+
+	// Evaluate all our possible annunciator sources. 
 
-#if defined(PIOS_INCLUDE_LED) && defined(PIOS_LED_HEARTBEAT)
-		// Flash the heartbeat LED
+	// The most important: indicate_error / alarms
 
-		FlightStatusData flightStatus;
-		FlightStatusGet(&flightStatus);
+	const char *candidate_blink;
+	uint8_t candidate_prio;
 
-		/* Quadruple heartbeat blink rate when armed */
-		unsigned int mask = flightStatus.Armed == FLIGHTSTATUS_ARMED_ARMED ?
-			1 : 7;
+	candidate_prio = indicate_error(&candidate_blink);
 
-		if (!(counter & mask)) {
-			PIOS_LED_Toggle(PIOS_LED_HEARTBEAT);
-			DEBUG_MSG("+ 0x%08x\r\n", 0xDEADBEEF);
+	if (candidate_prio > blink_prio) {
+		// Preempt!
+		blink_state = 0;
+		blink_string = candidate_blink;
+		blink_prio = candidate_prio;
+
+		if (blink_string &&
+				!strcmp(blink_string, BLINK_STRING_RADIO)) {
+			// XXX do this in a more robust way */
```

Comment:
Yah.. I think it's OK for now.  But if we're going to have fancier preemption behaviors, different alarms going different places, etc.  this whole section needs to change.